src/screens/SwapStack/swap-screen.tsx

```diff
@@ -0,0 +1,313 @@
+import {Loading} from '@app/components/ui';
+import {app} from '@app/contexts';
+import {AddressUtils} from '@app/helpers/address-utils';
+import {awaitForJsonRpcSign} from '@app/helpers/await-for-json-rpc-sign';
+import {awaitForValue} from '@app/helpers/await-for-value';
+import {getRpcProvider} from '@app/helpers/get-rpc-provider';
+import {useSumAmount, useTypedRoute} from '@app/hooks';
+import {Currencies} from '@app/models/currencies';
+import {SwapStackParamList, SwapStackRoutes} from '@app/route-types';
+import {Balance} from '@app/services/balance';
+import {
+  Indexer,
+  SushiPoolEstimateResponse,
+  SushiPoolResponse,
+} from '@app/services/indexer';
+import {IContract} from '@app/types';
+import {ERC20_ABI, V3SWAPROUTER_ABI} from '@app/variables/abi';
+import {
+  HAQQ_METADATA,
+  SUSHISWAP_MAINNET_ADDRESSES,
+  WETH_MAINNET_ADDRESS,
+} from '@app/variables/common';
+
+const logger = Logger.create('SwapScreen', {
+  emodjiPrefix: '🟠',
+  stringifyJson: __DEV__,
+});
+
+export const SwapScreen = observer(() => {
+  const {params} = useTypedRoute<SwapStackParamList, SwapStackRoutes.Swap>();
+  const [isEstimating, setIsEstimating] = useState(false);
+  const [isSwapInProgress, setSwapInProgress] = useState(false);
+  const [isApproveInProgress, setApproveInProgress] = useState(false);
+  const [tokenIn, setTokenIn] = useState<IContract | null>(null);
+  const [tokenOut, setTokenOut] = useState<IContract | null>(null);
+  const [estimateData, setEstimateData] =
+    useState<SushiPoolEstimateResponse | null>(null);
+  const amountsOut = useSumAmount(undefined, undefined, undefined, () => '');
+  const amountsIn = useSumAmount(
+    new Balance(1, 0),
+    undefined,
+    undefined,
+    () => '',
+  );
+  const [poolsData, setPoolsData] = useState<SushiPoolResponse>({
+    contracts: [],
+    pools: [],
+  });
+  const isLoading = useMemo(
+    () =>
+      !poolsData?.contracts?.length ||
+      !poolsData?.pools?.length ||
+      !tokenIn ||
+      !tokenOut,
+    [tokenIn, tokenOut, poolsData],
+  );
+
+  const estimate = useCallback(async () => {
+    try {
+      setIsEstimating(true);
+      const amount = new Balance(
+        parseFloat(amountsIn.amount),
+        tokenIn?.decimals!,
+        tokenIn?.symbol!,
+      );
+      logger.log('estimate', amount);
+      const response = await Indexer.instance.sushiPoolEstimate({
+        amount: amount.toHex(),
+        sender: params.address,
+        token_in: tokenIn?.eth_address!,
+        token_out: tokenOut?.eth_address!,
+        currency_id: Currencies.currency?.id,
+      });
+
+      logger.log('estimate', response);
+      setEstimateData(response);
+      const amountOut =
+        parseInt(response.amount_out, 16) / Math.pow(10, tokenOut?.decimals!);
+      amountsOut.setAmount(amountOut.toString());
+    } catch (err) {
+      logger.captureException(err, 'estimate');
+    } finally {
+      setIsEstimating(false);
+    }
+  }, [
+    params,
```

Comment:
I think you can skip memoization for this function, there is too many arguments for this

Reply:
I prefer to control the dependencies of useMemo/useCallback/useEffect for more precise behavior tuning. Moreover, this component has many other variables that didn’t make it into the dependencies.